File: 2021/04/04.js
let [nums, ...input] = require('fs').readFileSync('input', 'utf8').trim().split('\n\n').map(w => w.includes('\n') ? w.split('\n').map(y => y.split(' ').filter(z => z).map(z => +z)) : w.split(',').map(x => +x))

let check = board => board.some(row => row.reduce((a,b)=>a+b) == -5) || board.some((_,i) => board.reduce((a,b)=>a+b[i],0) == -5)
let pick = (bs, n) => bs.map(b => b.map(r => r.map(e => e == n ? -1 : e)))

console.log(nums.reduce(([bs, cc], n) => bs.length == 5 ? [bs,cc] : (bs.find(check) ? [bs.find(check), cc] : [pick(bs, n), n]), [input, -1]).reduce((a,b)=>a*(~~b?b:b.flat().filter(x => x != -1).reduce((a,b)=>a+b)),1))

console.log(nums.reduce(([bs, cc], n) => (bs.length == 1 ? (check(bs[0]) ? [bs, cc] : [pick(bs, n), n]) : [pick(bs, n).filter(x => !check(x)), n]), [input, -1]).reduce((a,b)=>a*(~~b?b:b[0].flat().filter(x => x != -1).reduce((a,b)=>a+b)),1))

File: 2021/04/04.exs
# input parsing
[nums | boards] = File.read!("input") |> String.split("\n\n")
nums = nums |> String.split(",") |> Enum.map(&String.to_integer/1)
boards =
  Enum.map(boards, fn board ->
    board
    |> String.split("\n")
    |> Enum.map(fn row ->
      row
      |> String.split(" ", trim: true)
      |> Enum.map(&String.to_integer/1)
    end)
  end)

# helper functions
mark = fn (boards, n) -> update_in(boards, [Access.all(), Access.all(), Access.all()], fn ^n -> nil; x -> x end) end
win? = fn board -> [board, Enum.map(Enum.zip(board), &Tuple.to_list/1)] |> Enum.any?(fn b -> Enum.any?(b, fn r -> Enum.all?(r, &is_nil/1) end) end) end
score = fn (board, n) -> board |> List.flatten() |> Enum.reject(&is_nil/1) |> Enum.sum() |> Kernel.*(n) end

# part 1
Enum.reduce_while(nums, boards, fn n, boards ->
  boards = mark.(boards, n)
  case Enum.find(boards, win?) do
    nil -> {:cont, boards}
    board -> {:halt, score.(board, n)}
  end
end)
|> IO.puts()

# part 2
Enum.reduce_while(nums, boards, fn
  n, [board] ->
    [board] = mark.([board], n)
    if win?.(board), do: {:halt, score.(board, n)}, else: {:cont, [board]}
  n, boards ->
    {:cont, boards |> mark.(n) |> Enum.reject(win?)}
end)
|> IO.puts()
